src/zenml/services/local/local_service_endpoint.py

```diff
@@ -0,0 +1,157 @@
+    concerning the life-cycle management and tracking of endpoints exposed
+    by external services implemented as local daemon processes.
+
+    Attributes:
+        config: service endpoint configuration
+        status: service endpoint status
+        monitor: optional service endpoint health monitor
+    """
+
+    config: LocalDaemonServiceEndpointConfig = Field(
+        default_factory=LocalDaemonServiceEndpointConfig
+    )
+    status: LocalDaemonServiceEndpointStatus = Field(
+        default_factory=LocalDaemonServiceEndpointStatus
+    )
+    monitor: Optional[
+        Union[HTTPEndpointHealthMonitor, TCPEndpointHealthMonitor]
+    ] = Field(..., discriminator="type")
+
+    @classmethod
+    def port_is_available(cls, port: int) -> bool:
+        """Check if a TCP port is available on the local machine.
+
+        Args:
+            port: TCP port number
+
+        Returns:
+            True if the port is available, False otherwise
+        """
+        with socket.socket(socket.AF_INET, socket.SOCK_STREAM) as sock:
+            try:
+                result = sock.connect_ex(("localhost", port))
+                if result == 0:
+                    logger.debug(
+                        f"TCP port check: port {port} is already taken."
+                    )
+                    return False
+
+                logger.debug(f"TCP port check: port {port} is available.")
+                return True
+            except OSError as e:
+                logger.debug(f"Error checking TCP port {port}: {str(e)}")
+                return False
+
+    def _lookup_free_port(self) -> int:
+        """Search for a free TCP port for the service endpoint.
+
+        If a preferred TCP port value is explicitly requested through the
+        endpoint configuration, it will be checked first. If a port was
+        previously used the last time the service was running (i.e. as
+        indicated in the service endpoint status), it will be checked next for
+        availability.
+
+        As a last resort, this call will search for a free TCP port, if
+        `allocate_port` is set to True in the endpoint configuration.
+
+        Returns:
+            An available TCP port number
+
+        Raises:
+            IOError: if the preferred TCP port is busy and `allocate_port` is
+            disabled in the endpoint configuration, or if no free TCP port
+            could be otherwise allocated.
+        """
+
+        # If a port value is explicitly configured, attempt to use it first
+        if self.config.port:
+            if self.port_is_available(self.config.port):
+                return self.config.port
+            if not self.config.allocate_port:
+                raise IOError(f"TCP port {self.config.port} is not available.")
+
+        # Attempt to reuse the port used when the services was last running
+        if self.status.port and self.port_is_available(self.status.port):
```

Comment:
Essentially status.port is the one that was last used and config.port is a new configuration? I'm a little confused by the port on both objects